# Exploring Ebay Car Sales Data


We'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
The aim of this project is to clean the data and make some basic analyzation of used car listings. 

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv",encoding="Windows-1252")

In [3]:
# Getting the basic overview of our DF
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
# Printing names of columns so we can decide if we should adjust them
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
# Editing columns' names from camelcase to snakecase and rewording some of the column names to be more descriptive.
new_col = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [6]:
# Checking the result
autos.columns = new_col
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   

   power_ps   model   odometer  registration_month fuel_type    brand  \
0       158  andere  150,000km                   3       lpg  peugeot   
1       286     7er  150,000km                   6    benzin      bmw   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

### __We will start exploring the data__

In [7]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

1. From the first look we can tell that columns "seller" and "offer_type" will not be useful because they both have frequency of 49999 for just one specific value (from 50 000 total amount). These can be dropped as they don't have useful information for analysis. 
2. Columns "price", "odometr", "name" will be cleaned.
3. Columns "power_ps", all columns with dates, "nr_of_pictures" and "postal_code" need more investigation. 

In [8]:
# After investigation it is shown that column "nr_of_pictures" has just zero values and columns with dates can be adjusted as well.
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

In [9]:
# We will drop three columns
autos.drop(labels= ["seller", "offer_type", "nr_of_pictures"], axis=1 ,inplace= True)
autos.head(1)

date_crawled                              name   price   abtest  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  $5,000  control   

  vehicle_type  registration_year  gearbox  power_ps   model   odometer  \
0          bus               2004  manuell       158  andere  150,000km   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

### __"price", "odometr" cleaning__

In [10]:
# Now we will work with numeric data stored as text which can be cleaned and converted.
# Firstly we will get rid of any non-numerical characters in columns "price", "odometr" and then convert their values into type integer
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km",'').astype(int)

In [11]:
# We will change original names of those two columns so that the new name would describe values
autos.rename(columns= {"price":"price_$", "odometer":"odometer_km"}, inplace=True)

#### __We will analyze the contents of both columns and look for outliers__


In [12]:
print("In column 'price_$' we have " + str(autos["price_$"].unique().shape) + "unique values and in 'odometer_km' " + str(autos["odometer_km"].unique().shape))

In column 'price_$' we have (2357,)unique values and in 'odometer_km' (13,)


In [13]:
autos["price_$"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

__We can see that 75% of all prices are no larger than 7 200$__

In [14]:
autos["price_$"].value_counts(ascending=False).head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price_$, dtype: int64

__The most common is price 0, which we can aprehand as an asking price, where the submitter waits for an offer and all the most common prices are quite low as well__

In [15]:
autos["price_$"].value_counts(ascending=True).head(5)

33980    1
2895     1
846      1
8970     1
20790    1
Name: price_$, dtype: int64

__There is a variety of uniquely priced cars__

In [16]:
autos["price_$"].quantile(q=[0.1,0.15,0.2,0.85,0.90,0.95])

0.10      500.0
0.15      699.0
0.20      900.0
0.85    10900.0
0.90    14000.0
0.95    19900.0
Name: price_$, dtype: float64

In [17]:
autos["price_$"].quantile(q=[0.96,0.97,0.98,0.99,1])

0.96       21950.0
0.97       24551.5
0.98       28000.0
0.99       35900.0
1.00    99999999.0
Name: price_$, dtype: float64

__We can see that as the price of the car grows larger, there is a larger and larger gap between them and even if we have a few more values which are larger than 20 000 and are in large variety 95% of all cars are priced below 19 900$__

In [18]:
print("5% of the total is " + str(len(autos["price_$"])*0.05) + " cars")

5% of the total is 2500.0 cars


__We would have to drop 2 500 cars as outliers which is still quit a lot. But from above percentiles we can see that jumps in prices between 95% and 99% are large but the largest jumps are in that one last percentage, so those will be the outliers__

In [19]:
autos["price_$"].quantile(q=[0.991,0.993,0.995,0.997,0.998,0.999,1])

0.991    3.780000e+04
0.993    4.090063e+04
0.995    4.700000e+04
0.997    5.780210e+04
0.998    6.990019e+04
0.999    1.050050e+05
1.000    1.000000e+08
Name: price_$, dtype: float64

In [20]:
print("All the prices in the last 0.1% are definitelly outlires  and we will get rid of " + str(len(autos["price_$"])*0.001) + " cars")

All the prices in the last 0.1% are definitelly outlires  and we will get rid of 50.0 cars


__We can presume that the last 0.1% within the highest prices can be classic cars, sports and exotic cars which may cost much more and are themselves outliers by nature or they can be just mistakes. On the other side there are a lot of very low values in this data and excluding them would have meant excluding a lot of the data, which would make for a weaker model. And we can interprete them as owners who are waiting for an offer rather than what was actually received.__

In [21]:
# Removing outliers
autos = autos[autos.loc[:,"price_$"].between(0,100000)]

#### __We will inspect the "odometer_km" column as well__

In [22]:
autos["odometer_km"].describe()

count     49947.000000
mean     125801.549643
std       39966.521087
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [23]:
autos["odometer_km"].value_counts()

150000    32413
125000     5167
100000     2166
90000      1757
80000      1435
70000      1230
60000      1162
50000      1023
5000        957
40000       817
30000       783
20000       776
10000       261
Name: odometer_km, dtype: int64

__Here we don't have to set any outliers because the amounts of cars which are being sold based on their odometr seems quite logical. The smallest amounts of cars are those with the lowest mileage. Simply said, they will last longer and there is no reason to sell them and at the opposite site, the biggest amount of cars being sold are those with loads of kilometres, so they can breake down any time and owners want to get rid of them.__

### __Inspecting columns containing dates.__

In [24]:
#We'll explore each of these columns to learn more about the listings.
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [25]:
# At the first sight we can tell that all the columns have written in the first ten characters the date, so we will work with just those.
# Then we will order dates from the oldest and display them with the amount of ads for those days in % (*100)
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025367
2016-03-06    0.013955
2016-03-07    0.035998
2016-03-08    0.033255
2016-03-09    0.033215
2016-03-10    0.032154
2016-03-11    0.032514
2016-03-12    0.036779
2016-03-13    0.015577
2016-03-14    0.036659
2016-03-15    0.033976
2016-03-16    0.029511
2016-03-17    0.031513
2016-03-18    0.013034
2016-03-19    0.034937
2016-03-20    0.037840
2016-03-21    0.037480
2016-03-22    0.032835
2016-03-23    0.032394
2016-03-24    0.029111
2016-03-25    0.031754
2016-03-26    0.032474
2016-03-27    0.031033
2016-03-28    0.034837
2016-03-29    0.034176
2016-03-30    0.033596
2016-03-31    0.031914
2016-04-01    0.033776
2016-04-02    0.035438
2016-04-03    0.038681
2016-04-04    0.036519
2016-04-05    0.013094
2016-04-06    0.003183
2016-04-07    0.001422
Name: date_crawled, dtype: float64

__Above we can see all dates when were data obtained (scraped from wepage to dataset) from the oldest data set from 2016-03-05 to the newest one from 2016-04-07. We can say that the most days the 'crawler' acquired slightly more than 3% from the total amount of data and it took him a month.__

In [26]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033255
2016-03-10    0.031894
2016-03-11    0.032815
2016-03-12    0.036619
2016-03-13    0.016938
2016-03-14    0.035257
2016-03-15    0.033736
2016-03-16    0.030012
2016-03-17    0.031173
2016-03-18    0.013715
2016-03-19    0.033876
2016-03-20    0.037880
2016-03-21 

__Here we can see that the oldest data about cars were added(the ad was created) on 2015-06-11 and that day was added 0,002% of the total amount. The newest set is from 2016-04-07 with 0.12% from the total.__

In [27]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001081
2016-03-06    0.004425
2016-03-07    0.005366
2016-03-08    0.007588
2016-03-09    0.009830
2016-03-10    0.010771
2016-03-11    0.012533
2016-03-12    0.023825
2016-03-13    0.008990
2016-03-14    0.012814
2016-03-15    0.015897
2016-03-16    0.016437
2016-03-17    0.027950
2016-03-18    0.007408
2016-03-19    0.015757
2016-03-20    0.020702
2016-03-21    0.020722
2016-03-22    0.021603
2016-03-23    0.018580
2016-03-24    0.019541
2016-03-25    0.019220
2016-03-26    0.016918
2016-03-27    0.016017
2016-03-28    0.020842
2016-03-29    0.022344
2016-03-30    0.024866
2016-03-31    0.023845
2016-04-01    0.023125
2016-04-02    0.024886
2016-04-03    0.025367
2016-04-04    0.024646
2016-04-05    0.124152
2016-04-06    0.221014
2016-04-07    0.130939
Name: last_seen, dtype: float64

__And the column "last_seen" shows us what amount of data was adjusted at certain dates. It is very similar to the column "date_crawled".__

#### __For the further use we will convert the dates to uniform numeric data__

In [28]:
# Keeping just the first ten characters and getting rid of "-"
autos["date_crawled"] = autos["date_crawled"].str[:10].str.replace("-","").astype(int)
autos["ad_created"] = autos["ad_created"].str[:10].str.replace("-","").astype(int)
autos["last_seen"] = autos["last_seen"].str[:10].str.replace("-","").astype(int)

#### __And the last interesting column with dates is "registration_year"__


In [29]:
autos["registration_year"].describe()

count    49947.0000
mean      2005.0755
std        105.7677
min       1000.0000
25%       1999.0000
50%       2003.0000
75%       2008.0000
max       9999.0000
Name: registration_year, dtype: float64

__Already we can tell that there are a lot of incorrect values, we can see that at minimum and maximum values.__

__We can say that maximum cannot be 9999 which is far into the future. The newest registration cannot be more than the year when the add was uploaded, which is 2016.__
__And the oldest year  should definitely not be before 1950, we will have a look on amounts of cars registered in specific years.__

In [30]:
# Getting amounts in % (*100) for each year from the oldest.
autos["registration_year"].value_counts(normalize=True, dropna=False).sort_index()

1000    0.000020
1001    0.000020
1111    0.000020
1500    0.000020
1800    0.000040
1910    0.000180
1927    0.000020
1929    0.000020
1931    0.000020
1934    0.000040
1937    0.000080
1938    0.000020
1939    0.000020
1941    0.000040
1943    0.000020
1948    0.000020
1950    0.000060
1951    0.000020
1952    0.000020
1953    0.000020
1954    0.000040
1955    0.000040
1956    0.000100
1957    0.000040
1958    0.000080
1959    0.000140
1960    0.000661
1961    0.000120
1962    0.000080
1963    0.000180
          ...   
2001    0.054077
2002    0.050714
2003    0.054598
2004    0.054798
2005    0.060364
2006    0.054197
2007    0.046109
2008    0.044627
2009    0.041985
2010    0.031894
2011    0.032695
2012    0.026428
2013    0.016057
2014    0.013274
2015    0.007908
2016    0.026228
2017    0.029071
2018    0.009830
2019    0.000060
2800    0.000020
4100    0.000020
4500    0.000020
4800    0.000020
5000    0.000080
5911    0.000020
6200    0.000020
8888    0.000020
9000    0.0000

In [31]:
# Testing how many % from the total would still remain if we would get rid of all the rows with registrations before 1980 and after 2016
test_1 = autos[autos.loc[:,"registration_year"].between(1980,2016)]
prct = str(len(test_1)/500)+ "%"
print(prct)

94.892%


__So if we get rid of all the cars registered before 1980 and after 2016 we will lose just slightly more than 5%.__

In [32]:
# Droping those rows
autos = autos[autos.loc[:,"registration_year"].between(1980,2016)]

In [33]:
autos["registration_year"].value_counts(normalize=True)

2000    0.070670
2005    0.063546
1999    0.063188
2004    0.057687
2003    0.057476
2006    0.057054
2001    0.056928
2002    0.053387
1998    0.051638
2007    0.048539
2008    0.046980
2009    0.044198
1997    0.042722
2011    0.034418
2010    0.033575
1996    0.030435
2012    0.027821
1995    0.027631
2016    0.027610
2013    0.016903
2014    0.013974
1994    0.013911
1993    0.009379
1990    0.008325
2015    0.008325
1992    0.008199
1991    0.007503
1989    0.003815
1988    0.002972
1985    0.002192
1980    0.002044
1986    0.001602
1987    0.001581
1983    0.001117
1984    0.001117
1982    0.000906
1981    0.000632
Name: registration_year, dtype: float64

__From our dataset we can tell that the most common year for registering car was 2000, when 7% out of the total were signed. And the lowest amounts were in 80s.__

### Small analysis of popularity and its dependance on price or mileage

In [34]:
autos.head(3)

date_crawled                                        name  price_$   abtest  \
0      20160326            Peugeot_807_160_NAVTECH_ON_BOARD     5000  control   
1      20160404  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik     8500  control   
2      20160326                  Volkswagen_Golf_1.6_United     8990     test   

  vehicle_type  registration_year    gearbox  power_ps   model  odometer_km  \
0          bus               2004    manuell       158  andere       150000   
1    limousine               1997  automatik       286     7er       150000   
2    limousine               2009    manuell       102    golf        70000   

   registration_month fuel_type       brand unrepaired_damage  ad_created  \
0                   3       lpg     peugeot              nein    20160326   
1                   6    benzin         bmw              nein    20160404   
2                   7    benzin  volkswagen              nein    20160326   

   postal_code  last_seen  
0        79588   20160406  
1        71034   20160406  
2        35394   20160406

In [35]:
# We will pick those brands which are represented by more than 1 percent of the total amount.
brand_agg = autos["brand"].value_counts(normalize=True) 
brand_agg = brand_agg[brand_agg > 0.01]

In [36]:
brand_agg

volkswagen       0.213190
bmw              0.110800
opel             0.108544
mercedes_benz    0.094697
audi             0.087384
ford             0.069321
renault          0.047802
peugeot          0.029866
fiat             0.025650
seat             0.018379
skoda            0.016145
mazda            0.015323
nissan           0.015281
smart            0.014079
citroen          0.013889
toyota           0.012583
Name: brand, dtype: float64

In [37]:
brand_names = brand_agg.index.tolist()

In [38]:
# How many brands are left out
l_o = autos["brand"].value_counts(normalize=True)
l_o = l_o[l_o < 0.01]
len(l_o)

24

__The most common (popular) cars are Volkswagens with 21% from the total, second is BMW with 11% and third Opel with 10%. The one percent which we have left out still contains 24 more brands.__

In [39]:
# Creating average prices for each brand
cars_price = {}
for name in brand_names:
    avg = autos.loc[autos["brand"] == name, "price_$"].mean()
    cars_price[name] = round(avg, 2)

In [40]:
cars_price

{'volkswagen': 5190.65,
 'bmw': 7968.24,
 'opel': 2853.66,
 'mercedes_benz': 8265.53,
 'audi': 9050.5,
 'ford': 3356.47,
 'renault': 2347.19,
 'peugeot': 3033.85,
 'fiat': 2642.33,
 'seat': 4301.42,
 'skoda': 6352.74,
 'mazda': 4010.77,
 'nissan': 4664.89,
 'smart': 3542.71,
 'citroen': 3600.77,
 'toyota': 5108.27}

__The highest average price has Audi, then Mercedes and the third is BMW and all three are in top five of the most advertised cars. Volkswagen which is the most advertised has medium average price 5190$, and the lowest price of 2347 has Renault which is seventh most offered car. But suprisingly the third most advertised car Opel has the third lowest average price, so  there's a distinct price gap.__

### __We will do the same for mileage/odometr__

In [41]:
# Creating average mileage for each brand
cars_odo = {}
for name in brand_names:
    avg = autos.loc[autos["brand"] == name, "odometer_km"].mean()
    cars_odo[name] = round(avg, 2)
cars_odo

{'volkswagen': 128933.76,
 'bmw': 132645.04,
 'opel': 129621.36,
 'mercedes_benz': 131234.14,
 'audi': 129324.65,
 'ford': 124984.8,
 'renault': 128428.13,
 'peugeot': 127191.25,
 'fiat': 117255.55,
 'seat': 121697.25,
 'skoda': 111253.26,
 'mazda': 124745.53,
 'nissan': 118572.41,
 'smart': 99595.81,
 'citroen': 119893.78,
 'toyota': 115753.77}

__We will transform these two dictionaries firstly into series and then join them in one dataframe.__

In [42]:
price_series = pd.Series(cars_price)
odo_series = pd.Series(cars_odo)
comb_cars = pd.DataFrame(price_series, columns=['average_price'])
comb_cars['average_mileage'] = odo_series
comb_cars.head(6)

average_price  average_mileage
volkswagen           5190.65        128933.76
bmw                  7968.24        132645.04
opel                 2853.66        129621.36
mercedes_benz        8265.53        131234.14
audi                 9050.50        129324.65
ford                 3356.47        124984.80

__The top 6 most advertised brands have quite high mileage,  but if we compare it to average price it doesn't show any larger mutual dependence on.__

### Changing values in DF

__A large amount of values in our dataset is in german, so we will try to convert them.__

In [43]:
# Creating a list of unique values
word_list = []
for column in autos.columns:
    if autos[column].dtype == "object" and column not in ["name", "brand"] :
        word_list += autos[column].unique().tolist()

In [44]:
word_list

['control',
 'test',
 'bus',
 'limousine',
 'kleinwagen',
 'kombi',
 nan,
 'coupe',
 'suv',
 'cabrio',
 'andere',
 'manuell',
 'automatik',
 nan,
 'andere',
 '7er',
 'golf',
 'fortwo',
 'focus',
 'voyager',
 'arosa',
 'megane',
 nan,
 'a3',
 'clio',
 'vectra',
 'scirocco',
 '3er',
 'a4',
 '911',
 'cooper',
 '5er',
 'polo',
 'e_klasse',
 '2_reihe',
 'c_klasse',
 'corsa',
 'mondeo',
 'altea',
 'a1',
 'twingo',
 'a_klasse',
 'cl',
 '3_reihe',
 's_klasse',
 'sandero',
 'passat',
 'primera',
 'wrangler',
 'a6',
 'transporter',
 'astra',
 'v40',
 'ibiza',
 'micra',
 '1er',
 'yaris',
 'colt',
 '6_reihe',
 '5_reihe',
 'corolla',
 'ka',
 'tigra',
 'punto',
 'vito',
 'cordoba',
 'galaxy',
 '100',
 'sharan',
 'octavia',
 'm_klasse',
 'lupo',
 'fiesta',
 'superb',
 'meriva',
 'c_max',
 'laguna',
 'touran',
 '1_reihe',
 'm_reihe',
 'touareg',
 'seicento',
 'avensis',
 'vivaro',
 'x_reihe',
 'ducato',
 'carnival',
 'boxster',
 'signum',
 'zafira',
 'rav',
 'a5',
 'beetle',
 'c_reihe',
 'phaeton',
 '

In [45]:
# Creating translation dictionary map
map_dict = {'kleinwagen':"small car", 'kombi':"station wagon", 'andere':"other", 'manuell':"manual", "automatik":"automatic", 'e_klasse': "e_class", '2_reihe': '2_series', 'c_klasse':'c_class', 'a_klasse':'a_class', 's_klasse':'s_class', '3_reihe': '3_series', '6_reihe':'6_series', '5_reihe':'5_series', 'm_klasse':'m_class', '1_reihe':'1_series', 'm_reihe':'m_series', 'c_reihe':'c_series', 'i_reihe':'i_series', 'mx_reihe':'mx_series', 'z_reihe':'z_series', 'cx_reihe':"cx_series", 'cr0_reihe':"cr_series", 'b_klasse':'b_class', 'xc_reihe':"xc_series", 'v_klasse':'v_class', '4_reihe':"4_series", 'rx_reihe':"rx_series",'g_klasse':'g_class', 'benzin':"gas", "elektro":"electro", "nein":"no", "ja":"yes", '7er':"7_series", '3er':"3_series", '5er':"5_series", '1er':"1_series", '6er':"6_series"}

In [46]:
# Now we will just add all the remaing values, which don't need translating
word_list_2 = []
for column in autos.columns:
    word_list_2 += autos[column].unique().tolist()
    
for word in word_list_2:
    if word not in map_dict:
        map_dict[word] = word

In [47]:
# We will use the new translation dictionary to change values in DF
for column in autos.columns:
    autos[column] = autos[column].map(map_dict)

autos.head()

date_crawled                                               name price_$  \
0     20160326                   Peugeot_807_160_NAVTECH_ON_BOARD    5000   
1     20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik    8500   
2     20160326                         Volkswagen_Golf_1.6_United    8990   
3     20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...    4350   
4     20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...    1350   

    abtest   vehicle_type registration_year    gearbox power_ps     model  \
0  control            bus              2004     manual      158     other   
1  control      limousine              1997  automatic      286  7_series   
2     test      limousine              2009     manual      102      golf   
3  control      small car              2007  automatic       71    fortwo   
4     test  station wagon              2003     manual        0     focus   

  odometer_km registration_month fuel_type       brand unrepaired_damage  \
0      150000                  3       lpg     peugeot                no   
1      150000                  6       gas         bmw                no   
2       70000                  7       gas  volkswagen                no   
3       70000                  6       gas       smart                no   
4      150000                  7       gas        ford                no   

  ad_created postal_code last_seen  
0   20160326       79588  20160406  
1   20160404       71034  20160406  
2   20160326       35394  20160406  
3   20160312       33729  20160315  
4   20160401       39218  20160401

### Cleaning the "name" column

In [48]:
# This column is mixture of brands, types and other personal description
# We will replace "-" with spaces, make everything lowercase and split in several new columns
autos["name"] = autos["name"].str.replace("_", " ").str.lower()
autos["brand"] = autos["brand"].str.replace("_", " ")
autos[["name","type","description"]] = autos["name"].str.split(n=2,expand=True) # brand will stay in "name", the rest will be split in two new columns "type" and "description"
autos

date_crawled        name price_$   abtest   vehicle_type  \
0         20160326     peugeot    5000  control            bus   
1         20160404         bmw    8500  control      limousine   
2         20160326  volkswagen    8990     test      limousine   
3         20160312       smart    4350  control      small car   
4         20160401        ford    1350     test  station wagon   
5         20160321    chrysler    7900     test            bus   
6         20160320          vw     300     test      limousine   
7         20160316        golf    1990  control      limousine   
8         20160322        seat     250     test            NaN   
9         20160316     renault     590  control            bus   
11        20160316    mercedes     350  control            NaN   
12        20160331       smart    5299  control      small car   
13        20160323        audi    1350  control      limousine   
14        20160323     renault    3999     test      small car   
15        20160401    corvette   18900     test          coupe   
16        20160316        opel     350     test  station wagon   
17        20160329  volkswagen    5500     test          coupe   
18        20160326   verkaufen     300  control            bus   
19        20160317       mazda    4150  control            suv   
20        20160305        audi    3500     test  station wagon   
21        20160306     porsche   41500     test         cabrio   
22        20160328        mini   25450  control         cabrio   
23        20160310     peugeot    7999  control            bus   
24        20160403         bmw   48500  control      limousine   
25        20160321        ford      90  control  station wagon   
26        20160403  volkswagen     777  control      small car   
27        20160327         hat       0  control            NaN   
28        20160319        mini    5250  control      small car   
29        20160402    mercedes    4999     test  station wagon   
30        20160314     peugeot      80     test      small car   
...            ...         ...     ...      ...            ...   
49970     20160321          c4   15800  control            bus   
49971     20160329      w.lupo     950     test      small car   
49972     20160326    mercedes    3300  control            bus   
49973     20160327    mercedes    6000  control         cabrio   
49974     20160320        golf       0  control         cabrio   
49975     20160327       honda    9700  control      small car   
49976     20160319        audi    5900     test  station wagon   
49977     20160331    mercedes    5500  control      limousine   
49978     20160404    mercedes     900  control      limousine   
49979     20160320  volkswagen   11000     test      small car   
49980     20160312        ford     400  control  station wagon   
49981     20160315        opel    2000  control  station wagon   
49982     20160329       skoda    1950  control      small car   
49983     20160306        ford     600     test      small car   
49984     20160331     student       0     test            NaN   
49985     20160402    verkaufe    1000  control            NaN   
49986     20160404    chrysler   15900  control      limousine   
49987     20160322        audi   21990  control      limousine   
49988     20160328         bmw    9550  control          coupe   
49989     20160311          vw     150     test      small car   
49990     20160321    mercedes   17500     test      limousine   
49991     20160306  kleinwagen     500  control            NaN   
49992     20160310        fiat    4800  control      small car   
49993     20160315        audi    1650  control      small car   
49994     20160322        audi    5000  control  station wagon   
49995     20160327        audi   24900  control      limousine   
49996     20160328        opel    1980  control         cabrio   
49997     20160402        fiat   13200     test         cabrio   
49998     20160308        audi   22900  contro

In [49]:
# With the new column "name" we can fill missing values in "brand"
# Creating another mapping dictionary, beacuse some values are not names and some are in different way than they should be
autos.loc[autos["brand"]== "sonstige autos", "name"].unique().tolist()

['corvette',
 'ssangyong',
 'ssanyong',
 'guenstiges',
 '3',
 'dodge',
 'kaufe',
 'proton',
 'gut',
 'wartburg',
 'werkaufen',
 'pontiac',
 'ich',
 'cadillac',
 'andere',
 'buick',
 'vw',
 'mg',
 'microcar',
 'verkaufe',
 'super',
 'abarth',
 'auto',
 'ferrari',
 'iveco',
 'maserati',
 'melkus',
 'suche',
 'autos',
 'alpina',
 'golf',
 'brauch',
 'brilliance',
 'fahrzeuglackierer',
 'reddbeat1976',
 'knutschkugel',
 'pfau',
 'sommerreifen',
 'gmc',
 'combi',
 'strandbuggy',
 'talbot',
 'elektroauto',
 'lexus',
 'mopedauto',
 'bentley',
 'barkas',
 'rote',
 'asia',
 'alles',
 'mgf',
 'formel',
 'strandbuggy/',
 'isuzu',
 'original',
 'lincoln',
 'wohnwagen',
 'opel',
 'chevrolet',
 'iphone6',
 'aixam',
 'kleines',
 'fahrzeug',
 'piaggio',
 'morgan',
 'hummer',
 'freewiel',
 '2er',
 'suchen',
 'moin',
 'saangyong',
 'hole',
 'mazda',
 'schrottreif?',
 'rolls',
 'iphone',
 'ac',
 'tausch',
 'transporter',
 '25',
 'sportsltquadcab4x4',
 'jede',
 'willys',
 'aston',
 'export',
 'wir',
 'pro

In [50]:
# Changing values in "name"
map_dict_2 = {'corvette':'corvette', "ssangyong":"ssangyong",'ssangyoug':"ssangyong", 'triumph':'triumph', 'caterham':'caterham', 'ssanyong':"ssangyong",'sang':"ssangyong", 'tesla':'tesla', 'dodge':'dodge', 'dogde':'dodge', 'proton':'proton', 'wartburg':'wartburg', 'pontiac':'pontiac', 'cadillac':'cadillac', 'vw':"volkswagen", 'mg':'mg', 'ferrari':'ferrari', 'iveco':'iveco', 'maserati':'maserati', "melkus":"melkus", "golf":"volkswagen", "brilliance":"brilliance", "talbot":"talbot",'lexus':'lexus','bentley':'bentley', 'barkas':'barkas', "isuzu":"isuzu", 'lincoln':'lincoln', 'opel':'opel', 'chevrolet':'chevrolet', "aixam":"aixam", 'piaggio':'piaggio', 'hummer':'hummer', 'morgan':'morgan','mazda':'mazda','rolls': "rolls royce", 'aston':'aston martin', 'lotus':'lotus', 'westfield':'westfield' }
autos["name"] = autos["name"].map(map_dict_2)

In [51]:
# Filling now the missing brands 
autos.loc[autos["brand"]== "sonstige autos", "brand"] = autos["name"]

In [52]:
# There are still 187 brands missing, but the information is not in any other columns
autos["brand"].isnull().sum(axis = 0)

187

In [53]:
autos["brand"].replace(np.nan, "Other brand",inplace=True)

In [54]:
# "brand" = "name"
autos.drop("name", axis=1 ,inplace= True)


In [55]:
# Rearranging columns
autos = autos[['brand', 'model', 'type', 'vehicle_type', 'price_$', 'odometer_km', 'abtest', 
       'gearbox', 'power_ps', 
       'fuel_type', 'unrepaired_damage',
       'postal_code','description','registration_year', 'registration_month', 'ad_created',  'last_seen', 'date_crawled']]

In [56]:
# Cleaned, renamed, translated dataframe
autos

brand     model      type   vehicle_type price_$ odometer_km  \
0            peugeot     other       807            bus    5000      150000   
1                bmw  7_series      740i      limousine    8500      150000   
2         volkswagen      golf      golf      limousine    8990       70000   
3              smart    fortwo     smart      small car    4350       70000   
4               ford     focus     focus  station wagon    1350      150000   
5           chrysler   voyager     grand            bus    7900      150000   
6         volkswagen      golf      golf      limousine     300      150000   
7         volkswagen      golf        iv      limousine    1990      150000   
8               seat     arosa     arosa            NaN     250      150000   
9            renault    megane    megane            bus     590      150000   
11     mercedes benz       NaN      a140            NaN     350      150000   
12             smart    fortwo     smart      small car    5299       50000   
13              audi        a3        a3      limousine    1350      150000   
14           renault      clio      clio      small car    3999      150000   
15          corvette       NaN        c3          coupe   18900       80000   
16              opel    vectra    vectra  station wagon     350      150000   
17        volkswagen  scirocco  scirocco          coupe    5500      150000   
18               bmw  3_series      mein            bus     300      150000   
19             mazda     other   tribute            suv    4150      150000   
20              audi        a4        a4  station wagon    3500      150000   
21           porsche       911       911         cabrio   41500      150000   
22              mini    cooper    cooper         cabrio   25450       10000   
23           peugeot       NaN     boxer            bus    7999      150000   
24               bmw  5_series      535i      limousine   48500       30000   
25              ford       NaN    escort  station wagon      90      150000   
26        volkswagen      polo      polo      small car     777      125000   
27              ford       NaN     einer            NaN       0      150000   
28              mini    cooper    cooper      small car    5250      150000   
29     mercedes benz   e_class      benz  station wagon    4999      150000   
30           peugeot  2_series       206      small car      80      150000   
...              ...       ...       ...            ...     ...         ...   
49970        citroen        c4     grand            bus   15800       60000   
49971     volkswagen      lupo       1.0      small car     950      150000   
49972  mercedes benz      vito      benz            bus    3300      150000   
49973  mercedes benz       slk      benz         cabrio    6000      150000   
49974     volkswagen      golf         1         cabrio       0      150000   
49975          honda      jazz      jazz      small car    9700      100000   
49976           audi        80        80  station wagon    5900      150000   
49977  mercedes benz   c_class      benz      limousine    5500      150000   
49978  mercedes benz   e_class      benz      limousine     900      150000   
49979     volkswagen      polo      polo      small car   11000       70000   
49980           ford    escort    escort  station wagon     400      125000   
49981           opel     astra     astra  station wagon    2000      150000   
49982          skoda     fabia     fabia      small car    1950       90000   
49983           ford     focus     focus      small car     600      150000   
49984    Other brand       NaN     sucht            NaN       0      150000   
49985     volkswagen       NaN    meinen            NaN    1000      150000   
49986       chrysler      300c      300c      limousine   15900      125000   
49987           audi        a3        a3      limousine   21990       50000   
49988            bmw  3_series       330          coupe    9550  